Support SmartStrip type devices (HS300, HS107) in tplink component

mattkasa opened the pull request:
Add support for SmartStrip type devices (HS300, HS107) to tplink component

## Description:

This adds a new `strip` type to the tplink component as well as detects strip type devices during discovery and uses a pyHS100 SmartStrip to iterate the SmartPlug(s) and adds each of them as a switch to HA. This also fixes state reporting for multi-plug devices and reports the state of each plug correctly.

**Related issue (if applicable):** fixes #25146

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10914

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

mattkasa committed:
Add support for SmartStrip type devices (HS300, HS107) to tplink component

rytilahti reviewed (commented):
Great to see someone willing to work to get this integrated! I left a couple of comments behind, it's been a while since I looked at this code (and there has been a major refactoring, which should improve potential problems), so you may want to check out how this would work with the most recent git master.

MartinHjelmare commented:
We use the dev branch as default branch. All PRs should be based off dev branch and target dev branch.

MartinHjelmare commented on homeassistant/components/tplink/switch.py:
```diff
@@ -122,8 +123,23 @@ def update(self):
+                self._state = (
+                    list(
+                        filter(
+                            lambda child: child["id"] == self.smartplug.context,
```
Please use a list comprehension instead of filter + lambda.

shbatm replied:
@mattkasa FWIW the following is working with this branch installed as a `custom_component`:

```python
                self._state = [
                    child
                    for child in self.smartplug.sys_info["children"]
                    if child["id"] == self.smartplug.context
                ][0]["state"] == 1
```

mattkasa replied:
Thank you for this @shbatm! I've incorporated this suggestion 👍

rytilahti commented on homeassistant/components/tplink/switch.py:
```diff
@@ -122,8 +123,23 @@ def update(self):
+                if self.smartplug.context is None:
+                    self._device_id = self.smartplug.sys_info["deviceId"]
+                else:
+                    self._device_id = self.smartplug.context
```
Hmm, I'm not sure what is going to happen for those who have been using this integration when the id suddenly changes, but I wouldn't mind to keep using context here and fix the issue upstream. But in case the device id changes, is there a need to migrate them? Can someone from @home-assistant/core clarify what's the best approach?

gjbadros replied:
This doesn't need to be a breaking change.  Without knowing about this change, I re-did a lot of this work (see doc change at https://github.com/home-assistant/home-assistant.io/pull/10382 and main PR at https://github.com/home-assistant/home-assistant/pull/26655 ).

To avoid breaking change, for SmartStrip Plug (and only for those plugs), give them a unique_id mac_address + "_" + devid (or append the plug number).  Non-smart-strip plugs can just use mac like before.

Note that I didn't notice that smartstrip builds a SmartPlug for each outlet (I was reverse engineering the sys_info and the network protocol and didn't even realize that the KP400 looks like a SmartStrip because I didn't know they made those).

shbatm replied:
Another FWIW: if you update this block (and remove the previous `self._alias` declaration) the individual outlets will use the names/aliases given to them in the Kasa app.

```python
                if self.smartplug.context is None:
                    self._device_id = self.smartplug.sys_info["deviceId"]
                    self._alias = self.smartplug.alias
                else:
                    self._device_id = self.smartplug.context
                    self._alias = [
                        child
                        for child in self.smartplug.sys_info["children"]
                        if child["id"] == self.smartplug.context
                    ][0]["alias"]
```

mattkasa replied:
Thanks again @shbatm, I've incorporated this with an additional change of setting `_device_id` to `_mac` for single plugs to retain backwards compatibility, so this will no longer will be a breaking change 👍

rytilahti commented on homeassistant/components/tplink/switch.py:
```diff
@@ -122,8 +123,23 @@ def update(self):
+                            self.smartplug.sys_info["children"],
+                        )
+                    )[0]["state"]
+                    == 1
```
Indeed, this was fixed later (in the current pyhs100 master) if I recall correctly. But until there will be a new release and the integration gets ported, I think it is fine to have it like it is even when it's kludgish.

mattkasa replied:
Sounds good 👍

MarkHofmann11 commented:
Curious if this PR is still being worked, as there is another (PR#26655) which adds similar support.  A blend of both would be perfect (to make a non-breaking change) and add the much requested support for TPlink power strips (or any multiple power outlet TPlink device).  KP400, HS300 and HS107.  Thanks everyone!

mattkasa commented:
I'll see if I can incorporate all the feedback into it :+1:

essell commented:
Very excited to see support for multi-outlet Kasa devices.  I've got an HS300 I can't wait to start using with Home Assistant. Thanks for all the hard work on supporting and extending the TP-Link integration.

mattkasa committed:
Merge branch 'dev' into tplink-multi-plug-support

mattkasa committed:
Incorporate feedback from @MartinHjelmare using changes suggested by @shbatm

- Setting `_state` now uses a list comprehension
- `_alias` will use aliases from the Kasa app
- `_device_id` will be set to `_mac` for single plugs to retain backwards compatibility

MartinHjelmare reviewed (approved):
Looks good!

@rytilahti?

MartinHjelmare commented:
Note that the coverage is failing, so we probably need to extend the tests before this can merge.

rytilahti commented:
Looks fine for me, the code can be cleaned up when the new pyhs100 gets released at some point. I don't have a test device so I cannot verify the functionality. Feel free to merge after the codecov check turns green.

My only concern is that this will cause a request for per socket, i.e., for six socket strip it will make 6 sysinfo calls even when the data would be accessible after the initial request. The current `master` of `pyHS100` has transparent caching for queries which will improve the situation, which is the reason why I mentioned it earlier.

mattkasa commented:
Makes sense, we'll definitely want to come back and refactor after the next pyhs100 release, no sense duplicating calls if we can help it. I'll work on getting coverage back up :+1:

rytilahti commented:
Ok, great :+1: After we get this merged in, I will start adapting the code to the new API and do some tests with my HS110 accordingly. I hope you don't mind if I ping you for some feedback & testing on the PR when it's time?

gjbadros commented:
Sure

On Mon, Oct 21, 2019, 10:06 AM Teemu R. <notifications@github.com> wrote:

> Ok, great 👍 After we get this merged in, I will start adapting the code
> to the new API and do some tests with my HS110 accordingly. I hope you
> don't mind if I ping you for some feedback & testing on the PR when it's
> time?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/26220?email_source=notifications&email_token=AALOHTIPC7ZYYYGBKIMYU43QPXOQFA5CNFSM4IPXJD2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB3B22A#issuecomment-544611688>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALOHTO737FZTWNOX7DN7FTQPXOQFANCNFSM4IPXJD2A>
> .
>

mattkasa commented:
Yep, absolutely! Happy to test with my HS107's any time!

gjbadros commented:
Just tested this against KP400s and and HS300 and it works as a custom
component (from 0.100.2, copied tplink and patched with this PR,
restarted).  I deleted the integration and restarted again as my old
devices were still there (but that could've been from my various hacks
experimenting with my PR and this one.

N.B. one feature request -- and maybe it's best done across the whole of
HASS or there's a trick I don't know about -- I'd love to see the IP
address in the extra information of the state view.   This is particularly
valuable for discovered components (although I avoid discovery for other
reasons).

Also note that the power draw reporting looks like it's working on the
HS300, but if it's only reported as an attribute, can you get graphs off of
that easily?  Should it create separate sensors for the power draw of each
plug or is that overkill?  Probably could use a template sensor off of the
attribute if one needs that.


Note t

On Mon, Oct 21, 2019 at 11:17 AM Greg Badros <badros@gmail.com> wrote:

> Sure
>
> On Mon, Oct 21, 2019, 10:06 AM Teemu R. <notifications@github.com> wrote:
>
>> Ok, great 👍 After we get this merged in, I will start adapting the code
>> to the new API and do some tests with my HS110 accordingly. I hope you
>> don't mind if I ping you for some feedback & testing on the PR when it's
>> time?
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/home-assistant/home-assistant/pull/26220?email_source=notifications&email_token=AALOHTIPC7ZYYYGBKIMYU43QPXOQFA5CNFSM4IPXJD2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB3B22A#issuecomment-544611688>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AALOHTO737FZTWNOX7DN7FTQPXOQFANCNFSM4IPXJD2A>
>> .
>>
>

MarkHofmann11 commented:
Awesome!!! :)  Not sure if there was an update already submitted for the tplink component documentation, but we should add the new strip option in the configuration.yaml settings.

tplink:
  discovery: false
  light:
   - host: x.x.x.x
  switch:
    - host: x.x.x.x
  dimmer:
    - host: x.x.x.x
  **strip:**
    - host: x.x.x.x

rytilahti commented:
I just merged the documentation update (linked in the PR description above), so I think we are all set. Thanks @mattkasa and @gjbadros for pushing this onwards!

@gjbadros your first question is more generic than just this integration, but I have no answer for it. For the second one, maybe the integration could be changed to create sensors for each device supporting that. At the moment the way is to use template sensors as shown in https://www.home-assistant.io/integrations/tplink#extracting-energy-sensor-data .

gjbadros commented:
@balloob re: the first question: I'd love to have a pair of standard attributes on all IP-based devices that are the mac address and the current IP address.  Integrations are inconsistent about how/if they make those properties available and without platform support, it won't happen consistently.

Re: my second question, yes, template sensors seem fine for grabbing energy data out of the devices from their attributes.

rytilahti commented:
@gjbadros the correct place to discuss on architectural changes is to create a proposal in https://github.com/home-assistant/architecture/ :-) But I would also personally like to see consolidation of such information on the platform/core level to make it consistent across the board (much like software version, hardware version, etc. is done).

gjbadros commented:
Thx for the pointer... when time permits I'll aim to make a proposal there
or at least start a discussion.  When a tplink or lifx device is
unavailable, without last-known-ip address and/or mac address, there's
really little further debugging one can do before just throwing up your
hands and restarting the server (or trying the underlying library, but
that's even more complex).

On Wed, Oct 23, 2019 at 4:20 PM Teemu R. <notifications@github.com> wrote:

> @gjbadros <https://github.com/gjbadros> the correct place to discuss on
> architectural changes is to create a proposal in
> https://github.com/home-assistant/architecture/ :-) But I would also
> personally like to see consolidation of such information on the
> platform/core level to make it consistent across the board (much like
> software version, hardware version, etc. is done).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/26220?email_source=notifications&email_token=AALOHTMG5JC6DWGRCLAGOZTQQDL25A5CNFSM4IPXJD2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECDFT6Q#issuecomment-545675770>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALOHTPAEOZILWJNDOTU573QQDL25ANCNFSM4IPXJD2A>
> .
>
